src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/ImmutableArrayExtensions.cs

```diff
@@ -205,4 +207,287 @@ public static ImmutableArray<T> GetMostRecentUniqueItems<T>(this ImmutableArray<
+        private readonly bool _useComparer;
+        private readonly bool _descending;
+
+        public CompareHelper(IComparer<T>? comparer, bool descending)
+        {
+            _comparerSpecified = comparer is not null;
+            _comparer = comparer ?? Comparer<T>.Default;
+            _useComparer = true;
+            _descending = descending;
+            _comparison = null!;
+        }
+
+        public CompareHelper(Comparison<T> comparison, bool descending)
+        {
+            _comparison = comparison;
+            _useComparer = false;
+            _descending = descending;
+            _comparer = null!;
+        }
+
+        public bool InSortedOrder(T? x, T? y)
+        {
+            // We assume that x and y are in sorted order if x is > y.
+            // We don't consider x == y to be sorted because the actual sor
+            // might not be stable, depending on T.
+
+            return _useComparer
+                ? !_descending ? _comparer.Compare(x!, y!) > 0 : _comparer.Compare(y!, x!) > 0
+                : !_descending ? _comparison(x!, y!) > 0 : _comparison(y!, x!) > 0;
+        }
+
+        public IComparer<T> GetOrCreateComparer()
+            // There are six cases to consider.
+            => (_useComparer, _comparerSpecified, _descending) switch
+            {
+                // Provided a comparer and the results are in ascending order.
+                (true, true, false) => _comparer,
+
+                // Provided a comparer and the results are in descending order.
+                (true, true, true) => DescendingComparer<T>.Create(_comparer),
+
+                // Not provided a comparer and the results are in ascending order.
+                // In this case, _comparer was already set to Comparer<T>.Default.
+                (true, false, false) => _comparer,
+
+                // Not provided a comparer and the results are in descending order.
+                (true, false, true) => DescendingComparer<T>.Default,
+
+                // Provided a comparison delegate and the results are in ascending order.
+                (false, _, false) => Comparer<T>.Create(_comparison),
+
+                // Provided a comparison delegate and the results are in descending order.
+                (false, _, true) => DescendingComparer<T>.Create(_comparison)
+            };
+    }
+
+    private abstract class DescendingComparer<T> : IComparer<T>
+    {
+        private static IComparer<T>? s_default;
+
+        public static IComparer<T> Default => s_default ??= new WrappedComparer(Comparer<T>.Default);
+
+        public static IComparer<T> Create(IComparer<T> comparer)
+            => new WrappedComparer(comparer);
+
+        public static IComparer<T> Create(Comparison<T> comparison)
+            => new WrappedComparison(comparison);
+
+        public abstract int Compare(T? x, T? y);
+
+        private sealed class WrappedComparer(IComparer<T> comparer) : DescendingComparer<T>
```

Comment:
Nit: `SwitchedComparer`, `FlippedComparer`? Obviously an implementation detail but wrapped doesn't suggest we're actually doing things with the args to me.

Reply:
That's totally reasonable